DataSource/Util/ChoicesBuilder.php

```diff
@@ -22,39 +23,71 @@ public function __construct(EntityManager $entityManager)
-                $repository = $this->getSpecifiedRepository($choicesConfiguration['repository']);
-            } else {
-                $repository = $this->entityManager->getRepository($choicesConfiguration['repository']);
-            }
-            $choicesField = $choicesConfiguration['field'];
-            $choicesSort = isset($choicesConfiguration['sort']) ? $choicesConfiguration['sort'] : null;
-            $choices = $this->getChoicesFromRepository($repository, $choicesField, $choicesSort);
-        } elseif (is_callable($choicesConfiguration)) {
-            // Choices is a callable. It will generate the choices array.
-            $choices = $choicesConfiguration();
-            if (!is_array($choices)) {
-                throw new \Exception("Choices callback defined in table configurations must return arrays");
-            }
-        } elseif (is_array($choicesConfiguration)) {
-            // Just a plain array. Use it as choices
-            $choices = $choicesConfiguration;
+            return $this->getChoicesFromRepository($choicesConfiguration);
+        }
+        if (is_callable($choicesConfiguration)) {
+            return $this->getChoicesFromCallable($choicesConfiguration);
+        }
+        if (is_array($choicesConfiguration)) {
+            return $choicesConfiguration;
         }
 
-        return $choices;
+        throw new \Exception('No usable configuration was found');
+    }
+
+    /**
+     * @param array $choicesConfiguration
+     *
+     * @return mixed
+     *
+     * @throws \Exception
+     */
+    private function getChoicesFromRepository(array $choicesConfiguration)
+    {
+        if ($choicesConfiguration['repository'] instanceof EntityRepository) {
+            $repository = $this->getSpecifiedRepository($choicesConfiguration['repository']);
+        } else {
+            $repository = $this->entityManager->getRepository($choicesConfiguration['repository']);
+        }
+
+        if (isset($choicesConfiguration['field'])) {
+            return $this->getChoicesFromRepositoryForField(
+                $repository,
+                $choicesConfiguration['field'],
+                isset($choicesConfiguration['sort']) ? $choicesConfiguration['sort'] : null
+            );
+        }
+        if (isset($choicesConfiguration['method'])) {
+            return $this->getChoicesFromRepositoryWithMethod(
+                $repository,
+                $choicesConfiguration['method']
+            );
+        }
+
+        throw new \Exception('Repository source expects field or method parameter');
+    }
+
+    /**
+     * @deprecated Repository should be specified by it's name, not as an actual object.
+     *
+     * @param EntityRepository $repository
+     *
+     * @return EntityRepository
+     */
+    private function getSpecifiedRepository(EntityRepository $repository)
+    {
+        trigger_error('Repository should be specified by it\'s name, not as an actual object.', E_USER_DEPRECATED);
```

Comment:
*Specifying repository as an object is deprecated and will not be supported since 1.0.0. Specify it by it's name instead.*